pkg/source/kafka/kafka_source.go

```diff
@@ -0,0 +1,361 @@
+	Assignor         string `hcl:"assignor,optional" env:"SOURCE_KAFKA_ASSIGNOR"`
+	TargetVersion    string `hcl:"target_version,optional" env:"SOURCE_KAFKA_SOURCE_VERSION"`
+	EnableSASL       bool   `hcl:"enable_sasl,optional" env:"SOURCE_KAFKA_ENABLE_SASL"`
+	SASLUsername     string `hcl:"sasl_username,optional" env:"SOURCE_KAFKA_SASL_USERNAME" `
+	SASLPassword     string `hcl:"sasl_password,optional" env:"SOURCE_KAFKA_SASL_PASSWORD"`
+	SASLAlgorithm    string `hcl:"sasl_algorithm,optional" env:"SOURCE_KAFKA_SASL_ALGORITHM"`
+	TLSCert          string `hcl:"tls_cert,optional" env:"SOURCE_KAFKA_TLS_CERT_B64"`
+	TLSKey           string `hcl:"tls_key,optional" env:"SOURCE_KAFKA_TLS_KEY_B64"`
+	TLSCa            string `hcl:"tls_ca,optional" env:"SOURCE_KAFKA_TLS_CA_B64"`
+	SkipVerifyTLS    bool   `hcl:"skip_verify_tls,optional" env:"SOURCE_KAFKA_TLS_SKIP_VERIFY_TLS"`
+}
+
+// kafkaSource holds a new client for reading messages from Apache Kafka
+type kafkaSource struct {
+	config           *sarama.Config
+	concurrentWrites int
+	topic            string
+	brokers          string
+	consumerName     string
+	log              *log.Entry
+	cancel           context.CancelFunc
+
+	client sarama.ConsumerGroup
+}
+
+// consumer represents a Sarama consumer group consumer
+type consumer struct {
+	concurrentWrites int
+	throttle         chan struct{}
+	source           *sourceiface.SourceFunctions
+	log              *log.Entry
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (consumer *consumer) Setup(sarama.ConsumerGroupSession) error {
+	consumer.log.Debugf("New session started")
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (consumer *consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	consumer.log.Debugf("Session ended, all ConsumeClaim goroutines exited")
+	return nil
+}
+
+// ConsumeClaim claims consumed messages and writes them to the target
+func (consumer *consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	wg := sync.WaitGroup{}
+	var consumeErr error
+	for message := range claim.Messages() {
+		wg.Add(1)
+		consumer.throttle <- struct{}{}
+		go func(message *sarama.ConsumerMessage) {
+			var messages []*models.Message
+			consumer.log.Debugf("Read message with key: %s", string(message.Key))
+
+			newMessage := &models.Message{
+				Data:         message.Value,
+				PartitionKey: uuid.NewV4().String(),
+				TimeCreated:  message.Timestamp,
+				TimePulled:   time.Now().UTC(),
+			}
+			if session != nil {
+				newMessage.AckFunc = func() {
+					consumer.log.Debugf("Ack'ing message with Key: %s", message.Key)
+					session.MarkMessage(message, "")
+				}
+			}
+
+			messages = append(messages, newMessage)
+
+			err := consumer.source.WriteToTarget(messages)
+			if err != nil {
+				consumer.log.Debugf("Error writing to target: %s", err)
+				consumeErr = err
```

Comment:
I think this should just be:

`consumer.log.WithFields(log.Fields{"error": err}).Error(err)`

When `WriteTotarget` returns an error it just means we failed to send some data - these messages won't have been acked so they'll get retried eventually. When we return it via `consumeErr`, it'll cause the app to crash and reboot, which is not the desired behaviour here.